### ISSUE TITLE
onScroll animation with useNativeDriver on iOS doesn't work after re-render

### ISSUE THREAD

**Issue by dgreensp** [facebook/react-native#18347]
Repro: https://snack.expo.io/SkkqUA4KG
Possibly same as: #12543

The conditions to reproduce seem to be:
* You have a ScrollView and are using an Animated.event onScroll
* You are on iOS with useNativeDriver: true
* A parent of a Animated.View that uses the animated contentOffset value re-renders (in this case, resulting in the child being unmounted and mounted again)

Expected result is that the Animated.View always animates based on the Animated.Value which is linked to the scroll position.

Actual result is that the scroll-linked animation works initially, but after the component that uses the animated value is destroyed and recreated, it doesn't work.  This only happens with the native driver.  It also doesn't happen if the same animated value is used continuously elsewhere (see comment in repro), so that could be a potential clue and workaround.

### Environment

```
Environment:
  OS: macOS High Sierra 10.13.3
  Node: 9.6.1
  Yarn: 1.5.1
  npm: 5.6.0
  Watchman: Not Found
  Xcode: Xcode 9.1 Build version 9B55
  Android Studio: Not Found

Packages: (wanted => installed)
  react: 16.2.0 => 16.2.0
  react-native: https://github.com/expo/react-native/archive/sdk-25.0.0.tar.gz => 0.52.0
```

**Comment by dgreensp** [comment]
Are there known gotchas about sharing Animated.Values between components?  If component A gives Animated.Value X to component B, and then B goes away and A gives X to C, is there some kind of teardown that happened, and now X is "dead"?

**Comment by dgreensp** [comment]
So far, I've learned that if you are in the buggy state, where onScroll isn't animating, anything that causes `_detachNativeEvents()` and `_attachNativeEvents()` to be called on the ScrollView will give it the "kick" it needs to start working again, including calling `forceUpdate()` on the ScrollView, a re-render of the parent, etc.

**Comment by dgreensp** [comment]
Aha!  I think I have an idea what's going on.  A bug in the native animation driver.

Background (gleaned from reading the source): All AnimatedValues driven by the native driver get marked as "native" AnimatedValues.  Native AnimatedValues have "native tags" which are id numbers used to look up nodes in the value graph on the native side of things, but they only have these tags when "attached."  AnimatedValues become attached when they are consumed, directly or transitively, such as by an Animated.View as a prop or style prop, and are detached when they stop being used.  When they start being used again, they are attached again, but with a fresh native tag.

So, my theory is that Animated.event is setting up drivers in native land that identify AnimatedValues by their native tag, and when that native tag changes, the drivers stop working.

A workaround is to make sure the AnimatedValues that are the target of the Animated.event are never detached, which can be done by passing them as pretty much any prop to any component, e.g. `dummyProp={myAnimatedValue}` on your ScrollView. :)  As far as I can tell, this doesn't result in any extra work being done, it just keeps the child count of your AnimatedValue from going to zero, so it stays in memory in native land and has a stable native tag.

**Comment by ssssssssssss** [comment]
@dgreensp Could you share some code snippets to workaround this issue? Thanks

**Comment by dgreensp** [comment]
Sure, glad to.  While it seemed to work at first to assign the AnimatedValue in question (the global variable `SCROLL` in the repro) to any old prop on a component (the ScrollView being a convenient component to slap it on, because then the workaround can live inside my scrolling component), when I tried applying this workaround in my actual app, I got an invariant exception saying an animation node didn't have a native tag.  For some reason, driving a component's props but not style with a native driver caused the problem.

Here's the workaround I ended up with, using the repro as context for this snippet:

```
export default class App extends Component {
  onScroll = Animated.event(
    [{ nativeEvent: { contentOffset: { y: SCROLL } } }],
    {useNativeDriver: true}
  );

  render() {
    return <View style={{flex: 1}}>
      <Animated.ScrollView
        style={{
          flex: 1,
          opacity: Animated.add(1, Animated.multiply(0, SCROLL))
        }}
        scrollEventThrottle={1}
        onScroll={this.onScroll}
        >
```

Basically, you make `opacity` or `transform` depend on whatever AnimatedValues you want to protect from being detached.

**Comment by ssssssssssss** [comment]
We actually use ```const AnimatedFlatList = Animated.createAnimatedComponent(FlatList)``` and attach onScroll prop with 
```           
          onScroll={Animated.event(                                   
             [{ nativeEvent: { contentOffset: { y: this.state.scrollY } } }],
             {                                                         
               useNativeDriver: true,                                  
               isInteraction: false
             } 
           )}
```
so it will do the trick if set an opacity style like you do, right?
```
        style={{
          flex: 1,
          // this line is the workaround:
          opacity: Animated.add(1, Animated.multiply(0, this.state.scrollY))
        }}
```

**Comment by dgreensp** [comment]
Yup, give it a shot

**Comment by bdilly** [comment]
I was able to reproduce this issue with RN 0.55.4 on Android

**Comment by iamralpht** [comment]
I ran into something similar which I think is the same root cause. In my case, my Animated.Values tracking scroll position are used by some other views which can be created, mounted & unmounted many times in the life time of the scroll view. After the first unmount, `__detach` is called on the scroll Animated.Values and they don't work properly after that (the JS value is out of sync with the native value, and change events no longer get through).

I can work around this by calling `__addChild` on the scroll Animated.Values (and `__removeChild` when my scroll view unmounts). The proposed workaround is essentially doing the same thing.

I think that a reasonable fix would be for `Animated.event` to become a child of the values it populates while it's attached to an event emitter. I don't think this would introduce cycles into the Animated.Value graph, but I'm not sure...
